### PR TITLE
handle parallel RollingUpgrades of multiple node groups

### LINKED ISSUE #57: node group RU can fail with a validation error
**BUG REPORT**

**What happened**:
During a parallelized (kiam and system at the same time rather than in sequence) nodegroup rollingUpgrade process, I'm seeing a common (most attempts) scenario where the Upgrade Manager Controller changes a node to stand-by. Then the node running this controller is drained (to be swapped out, itself). The controller is immediately re-spawned on another node and begins processing the rollingupgrades it knows about.

It finds the ongoing upgrade and once again attempts to move the nodegroup's selected node to stand-by. In some cases, this is effectively a no-op and is ignored. But in the failure case, it is instead flagged as an ValidationError. This outcome hints that the node may not have already been in stand-by nor was it still in in-service.

This can happen for any of the nodes in the sequence of a nodegroup and for either of the nodegroups (system or kiam).

Mostly, it seems to be a timing issue related to the controller re-starting and re-requesting stand-by for an already stand-by node. In case the node is in a different state than stand-by or in-service (a transitional state, ex: entering-stand-by, progressing), that could explain what I'm seeing.

**What you expected to happen**:
Redundant/follow-up stand-by requests should be a no-op. No validation error should occur. RU should continue as expected.

**How to reproduce it (as minimally and precisely as possible)**:
If this is a timing issue/race-condtion... it may be possible even in a sequential NodeGroup upgrade process. But it certainly seems to be exacerbated by parallel, interlaced upgrades.

See the console logs below to indicate the process which leads to this condition. They are long and rather dense in order to capture full sequence and requisite details to highlight scenario. Look for `//NOTE: ...` and `<...>` lines to provide context of what is happening throughout.

```
time="2020-02-21T21:59:27Z" level=info msg="selected instances for rotation: [{\n  AvailabilityZone: \"us-west-2a\",\n  HealthStatus: \"Healthy\",\n  InstanceId: \"i-09a13a158d5498ead\",\n  LifecycleState: \"InService\",\n  ProtectedFromScaleIn: false\n}]"
time="2020-02-21T21:59:27Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Found instance i-09a13a158d5498ead, name ip-10-198-55-69.us-west-2.compute.internal\n"
time="2020-02-21T21:59:27Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Found instance i-09a13a158d5498ead, name ip-10-198-55-69.us-west-2.compute.internal\n"
time="2020-02-21T21:59:27Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: setting instance i-09a13a158d5498ead state to in-progress"
time="2020-02-21T21:59:27Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: starting replacement with eager mode"

//NOTE: kiam node stand-by requested
time="2020-02-21T21:59:27Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Setting i-09a13a158d5498ead to stand-by"
time="2020-02-21T21:59:28Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: new instance has not yet joined the scaling group"
time="2020-02-21T21:59:29Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: node ip-10-198-57-152.us-west-2.compute.internal is unjoined from cluster, upgrade will proceed"
time="2020-02-21T21:59:29Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: setting instance i-05394222fa272dc32 state to completed"
time="2020-02-21T21:59:29Z" level=info msg="selected instances for rotation: [{\n  AvailabilityZone: \"us-west-2a\",\n  HealthStatus: \"Healthy\",\n  InstanceId: \"i-063ed0f802078438e\",\n  LifecycleState: \"InService\",\n  ProtectedFromScaleIn: false\n}]"
time="2020-02-21T21:59:29Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: Found instance i-063ed0f802078438e, name ip-10-198-55-228.us-west-2.compute.internal\n"
time="2020-02-21T21:59:29Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: Found instance i-063ed0f802078438e, name ip-10-198-55-228.us-west-2.compute.internal\n"
time="2020-02-21T21:59:29Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: setting instance i-063ed0f802078438e state to in-progress"
time="2020-02-21T21:59:29Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: starting replacement with eager mode"

//NOTE: system node stand-by requested
time="2020-02-21T21:59:29Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: Setting i-063ed0f802078438e to stand-by"
time="2020-02-21T21:59:29Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: new instance has not yet joined the scaling group"
time="2020-02-21T22:00:58Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: desired capacity is met, 3 instances in service"
time="2020-02-21T22:00:58Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: desired capacity is met, 3 nodes in service"
time="2020-02-21T22:00:58Z" level=info msg="Creating a context with timeout - 600"
time="2020-02-21T22:00:58Z" level=info msg="Invoking kubectl drain for the node - ip-10-198-55-69.us-west-2.compute.internal"
time="2020-02-21T22:00:58Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Running script /usr/local/bin/kubectl drain ip-10-198-55-69.us-west-2.compute.internal --ignore-daemonsets=true --delete-local-data=true --force --grace-period=-1"
time="2020-02-21T22:00:59Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: desired capacity is met, 3 instances in service"
time="2020-02-21T22:00:59Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: desired capacity is met, 3 nodes in service"
time="2020-02-21T22:00:59Z" level=info msg="Creating a context with timeout - 600"
time="2020-02-21T22:00:59Z" level=info msg="Invoking kubectl drain for the node - ip-10-198-55-228.us-west-2.compute.internal"
time="2020-02-21T22:00:59Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: Running script /usr/local/bin/kubectl drain ip-10-198-55-228.us-west-2.compute.internal --ignore-daemonsets=true --delete-local-data=true --force --grace-period=-1"
time="2020-02-21T22:01:00Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Script finished with output: node/ip-10-198-55-69.us-west-2.compute.internal cordoned\nWARNING: ignoring DaemonSet-managed Pods: kube-system/aws-node-vxc8l, kube-system/calico-node-td8kh, kube-system/kube-proxy-cjld2\nevicting pod \"kiam-server-5b578ff6dc-mrxw4\"\nevicting pod \"kiam-agent-all-nodes-healthcheck-workflow-28mj5-2557227070\"\npod/kiam-server-5b578ff6dc-mrxw4 evicted\npod/kiam-agent-all-nodes-healthcheck-workflow-28mj5-2557227070 evicted\nnode/ip-10-198-55-69.us-west-2.compute.internal evicted\n"
time="2020-02-21T22:01:00Z" level=info msg="Kubectl drain completed for node - ip-10-198-55-69.us-west-2.compute.internal"
time="2020-02-21T22:01:00Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Waiting for postDrainDelay of 90 seconds"
rpc error: code = Unknown desc = Error: No such container: 71c4dae0dd6bd2c5378bf076b319d073bc7d29eecc724a66e592ec59a161da4e

//NOTE: controller just stopped on node it was running on, migrating to new node
<commands to re-attach to logs of newly spun-up controller pod>

2020-02-21T22:01:11.375Z        INFO    setup   Watch RollingUpgrade objects only in all namespaces
2020-02-21T22:01:12.910Z        INFO    controller-runtime.metrics      metrics server is starting to listen    {"addr": ":8080"}
time="2020-02-21T22:01:12Z" level=info msg="max parallel reconciles = 10"
2020-02-21T22:01:12.912Z        INFO    setup   starting manager
2020-02-21T22:01:12.913Z        INFO    controller-runtime.controller   Starting EventSource    {"controller": "rollingupgrade", "source": "kind source: /, Kind="}
2020-02-21T22:01:12.913Z        INFO    controller-runtime.controller   Starting Controller     {"controller": "rollingupgrade"}
2020-02-21T22:01:12.913Z        INFO    controller-runtime.manager      starting metrics server {"path": "/metrics"}
2020-02-21T22:01:13.013Z        INFO    controller-runtime.controller   Starting workers        {"controller": "rollingupgrade", "worker count": 10}

<controller recognizing all registered rolling upgrades and determining which are not complete>

2020-02-21T22:01:13.019Z        INFO    controllers.RollingUpgrade      No more processing      {"rollingupgrade": "dmasselink-deleteme-kiam-ng-nodelaunchconfig-1b9gglwhwuyre", "Object state": "completed"}
2020-02-21T22:01:13.019Z        INFO    controllers.RollingUpgrade      Deleted object from admission map       {"rollingupgrade": "dmasselink-deleteme-kiam-ng-nodelaunchconfig-1b9gglwhwuyre"}
2020-02-21T22:01:13.019Z        INFO    controllers.RollingUpgrade      Adding obj to map:      {"rollingupgrade": "addon-upgrademgr-ns/dmasselink-deleteme-kiam-ng-nodelaunchconfig-1b9gglwhwuyre", "name": "dmasselink-deleteme-kiam-ng-nodelaunchconfig-1b9gglwhwuyre"}
2020-02-21T22:01:13.019Z        DEBUG   controller-runtime.controller   Successfully Reconciled {"controller": "rollingupgrade", "request": "addon-upgrademgr-ns/dmasselink-deleteme-kiam-ng-nodelaunchconfig-1b9gglwhwuyre"}
time="2020-02-21T22:01:13Z" level=info msg="Nodes in ASG dmasselink-deleteme-system-ng-NodeGroup-14BC0XSYA4B10 that *might* need to be updated: 4\n"
2020/02/21 22:01:13 Max unavailable calculated for dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8 is 1
time="2020-02-21T22:01:13Z" level=info msg="Nodes in ASG dmasselink-deleteme-kiam-ng-NodeGroup-1WV2HTDAY49SK that *might* need to be updated: 4\n"
2020/02/21 22:01:13 Max unavailable calculated for dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv is 1

//NOTE: kiam node is now in stand-by
time="2020-02-21T22:01:13Z" level=info msg="found in progress instances: [{\n  AvailabilityZone: \"us-west-2a\",\n  HealthStatus: \"Healthy\",\n  InstanceId: \"i-09a13a158d5498ead\",\n  LifecycleState: \"Standby\",\n  ProtectedFromScaleIn: false\n}]"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Found instance i-09a13a158d5498ead, name ip-10-198-55-69.us-west-2.compute.internal\n"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Found instance i-09a13a158d5498ead, name ip-10-198-55-69.us-west-2.compute.internal\n"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: setting instance i-09a13a158d5498ead state to in-progress"

//NOTE: system node is now in stand-by
time="2020-02-21T22:01:13Z" level=info msg="found in progress instances: [{\n  AvailabilityZone: \"us-west-2a\",\n  HealthStatus: \"Healthy\",\n  InstanceId: \"i-063ed0f802078438e\",\n  LifecycleState: \"Standby\",\n  ProtectedFromScaleIn: false\n}]"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: Found instance i-063ed0f802078438e, name ip-10-198-55-228.us-west-2.compute.internal\n"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: Found instance i-063ed0f802078438e, name ip-10-198-55-228.us-west-2.compute.internal\n"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: setting instance i-063ed0f802078438e state to in-progress"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: starting replacement with eager mode"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: Setting i-063ed0f802078438e to stand-by"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: starting replacement with eager mode"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Setting i-09a13a158d5498ead to stand-by"

//NOTE: kiam node's attempted re-stand-by treated as no-op (already in stand-by)
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: cannot set instance i-09a13a158d5498ead to stand-by, instance is already in stand-by"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: desired capacity is met, 3 instances in service"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: desired capacity is met, 3 nodes in service"
time="2020-02-21T22:01:13Z" level=info msg="Creating a context with timeout - 600"
time="2020-02-21T22:01:13Z" level=info msg="Invoking kubectl drain for the node - ip-10-198-55-69.us-west-2.compute.internal"
time="2020-02-21T22:01:13Z" level=info msg="dmasselink-deleteme-kiam-ng-nodelaunchconfig-kohevjehkijv: Running script /usr/local/bin/kubectl drain ip-10-198-55-69.us-west-2.compute.internal --ignore-daemonsets=true --delete-local-data=true --force --grace-period=-1"

//NOTE: system node's attempted re-stand-by treated as ValidationError (was it in another state by now?)
time="2020-02-21T22:01:14Z" level=error msg="dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8: instance i-063ed0f802078438e failed to enter standby: ValidationError: The instance i-063ed0f802078438e is not in InService.\n\tstatus code: 400, request id: 151b45aa-afe0-45ce-ac4c-68c4525dce9b"
time="2020-02-21T22:01:14Z" level=info msg="Deleted the entries of ASG dmasselink-deleteme-system-ng-NodeGroup-14BC0XSYA4B10 in the cluster store for dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8"
time="2020-02-21T22:01:14Z" level=info msg="Marked object dmasselink-deleteme-system-ng-nodelaunchconfig-1arm9ihu1urd8 as error"
```

### PR DESCRIPTION
**Pull request by davemasselink**
fixes #57

Overview
======
This PR will fix a subtle, but rather meaningful, bug in the UM codebase related to how the reconciler handles ASGs which led to the race/timing issue at the core of #57.

Testing
=====
After making this change, I've been able to successfully carryout >5 rolling upgrades of parallel nodegroups, without incident, in a row. Previously, I was seeing failures roughly 5 out of 6 runs.

Further, added a new unit test which ensures that two RUs can be considered one after the other, and that the correct ASG will be referenced each time.

### CODE REVIEW

File: api/v1alpha1/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -15,7 +15,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Code generated by controller-gen. DO NOT EDIT.
+// autogenerated by controller-gen object, do not modify manually
```
Comment (shrinandj) [inline api/v1alpha1/zz_generated.deepcopy.go]: Can you check the version of `controller-gen` on your local machine. Looks like you might be using a different one (probably older).

```
$ controller-gen --version
Version: v0.2.4
```

If you use the correct version as above, these spurious changes will go away.

**Comment by eytan-avisror** [comment]
Can you explain how this change fixes the bug? I am not sure why we need to have all ASGs in context of a single reconcile

**Comment by eytan-avisror** [comment]
@davemasselink can you try running this a few more times? I don’t think this will actually fix the bug. Since this bug is timing related it’s possible that it just worked in a few times you tried it.
A single reconcile thread has no need to know of other reconciles, hence this change is a no-op.
Please let me know if I am missing something 👍🏻

File: api/v1alpha1/zz_generated.deepcopy.go (hunk 1 above)
Comment (davemasselink) [reply api/v1alpha1/zz_generated.deepcopy.go]: yeah, you're right... had been investigating newer and older kubebuilder versions with Ravi recently re: another issue and I think I ended up building and linking an older version of controller-tools source
 
not sure what version I was on... it was before they even added the `--version` flag!

anyway updated binary and then quashed some incorrectly generated code/spec in latest commit

**Comment by eytan-avisror** [comment]
@davemasselink I take it back.
I just looked at the code and your changes again trying to understand better.
I see the bug this fixes now, `RollingUpgradeReconciler` is shared between reconciles, hence if we are storing a single ASG within it, it's a major problem for parallel reconciles, this context will only hold the last CR that was reconciled and exposes this race condition.

This becomes a problem mainly in controller restarts or when CRs are submitted during a existing rollups, which is the observed behavior.

Your change makes sense to me now, however we should use a thread safe map as it is accessed concurrently, e.g. `sync.Map` instead of `map[string]*autoscaling.Group`.

Thanks for catching this

File: controllers/rollingupgrade_controller.go [hunk 2]
```diff
@@ -112,7 +110,7 @@ type RollingUpgradeReconciler struct {
 	EC2Client       ec2iface.EC2API
 	ASGClient       autoscalingiface.AutoScalingAPI
 	generatedClient *kubernetes.Clientset
-	Asg             *autoscaling.Group
+	Asgs            map[string]*autoscaling.Group
```
Comment (eytan-avisror) [inline controllers/rollingupgrade_controller.go]: I wonder if we can get rid of this altogether and switch everything to using `ruObjNameToASG` which is already a sync.Map.
We just need to make sure we overwriting objects when we call `populateAsg`

File: controllers/rollingupgrade_controller.go [hunk 3]
```diff
@@ -262,8 +260,8 @@ func (r *RollingUpgradeReconciler) WaitForDesiredInstances(ruObj *upgrademgrv1al
 			return err
 		}
 
-		inServiceCount := getInServiceCount(r.Asg.Instances)
-		if inServiceCount == aws.Int64Value(r.Asg.DesiredCapacity) {
+		inServiceCount := getInServiceCount(r.Asgs[ruObj.Name].Instances)
```
Comment (eytan-avisror) [inline controllers/rollingupgrade_controller.go]: Potentially, if we make sure `ruObjNameToASG` gets updated by `populateAsg`, we can then load, cast and use it that way.

Example
```
	value, ok := r.ruObjNameToASG.Load(ruObj.Name)
	if !ok {
		msg := "Failed to find rollingUpgrade name in map."
		log.Printf(msg)
		return 0, errors.New(msg)
	}

	asg := value.(*autoscaling.Group)
```

and then
```
inServiceCount := getInServiceCount(asg.Instances)
```

File: controllers/rollingupgrade_controller.go [hunk 4]
```diff
@@ -464,11 +461,15 @@ func (r *RollingUpgradeReconciler) populateAsg(ruObj *upgrademgrv1alpha1.Rolling
...
 	}
 
 	asg := result.AutoScalingGroups[0]
-	r.Asg = asg
+	if r.Asgs == nil {
+		// Init map to hold all parallel reconciling ASGs
+		r.Asgs = make(map[string]*autoscaling.Group)
+	}
+	r.Asgs[ruObj.Name] = asg
 	r.ruObjNameToASG.Store(ruObj.Name, asg)
```
Comment (eytan-avisror) [inline controllers/rollingupgrade_controller.go]: I see it already gets saved here.
Lets just get rid of `r.Asgs` it is not needed at all when we store it everything in a sync.Map `ruObjNameToASG`

**Comment by davemasselink** [comment]
@eytan-avisror, thanks for the careful and skeptical read!

yes, ensuring that we're using a thread-safe data structure makes sense and seems critical, as you suggest

I like the idea of further relying on the `sync.Map`s already in-place. I'll update this code as you hint at and will add to the PR's description some more details about my testing strategy

File: controllers/rollingupgrade_controller.go (hunk 4 above)
Comment (davemasselink) [reply controllers/rollingupgrade_controller.go]: yes, good point! I'll change all uses of `r.Asg` (`r.Asgs` after my changes) to leverage this store you're highlighting

**Pull request merged**